---
title: "(1) Single-cell RNA seq pipeline - Preprocessing"
author: "Stefan Filges"
date: '`r format(Sys.Date())`'
output: 
  html_document:
    theme: sandstone
    highlight: tango
    code_folding: show
    toc: yes
    toc_depth: 3
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# About this document

This document provides the main single cell analysis workflow of

- loading single RNA-seq generated with 10x Genomics
- filtering, normalisation
- (optional) cell cycle regression
- dimensionality reduction
- clustering

The output are various plots and a data object containing the raw and processed
information for use in other applications.

Other scripts can be run on the data output from this document to do the following:

- differential gene expression analysis
- gene set enrichment analysis
- pseudotime ordering

Parameters at most steps can be altered but might change the results substantially. A attempt
was made to follow current best practices (as of `r format(Sys.Date())`), outlined below.

# Seurat pipeline
## Background and sources
Single-cell RNA-seq methods develop rapidly and no standardized workflow exists.
This analysis is based on current best practices as outlined by [Luecken and Theis](https://www.embopress.org/doi/10.15252/msb.20188746) (June 2019). Major analysis platforms
are Seurat and Scater (written in R) and scanpy (written in Python). A web-based
course and overview of scRNA-seq methods, including Seurat, is maintained by
[the Hemberg lab](https://scrnaseq-course.cog.sanger.ac.uk/website/index.html).

We use the Seurat package maintained by the [Satija lab](https://satijalab.org/seurat/) and available through [CRAN](https://cran.r-project.org/web/packages/Seurat/index.html). The version used here is
Seurat 3.0 (Released April 16, 2019). See their guided tutorials and the review
above for detailed explanations. Packages and algorithms used are linked to
published articles or repositories for further information.

## Loading packages and setting main parameters

First we load packages required for the main analysis, plotting and for later
pseudotime ordering using multiple different algorithms. 

At this step we also decide if cell cycle affects should be regressed using the
filter_cell_cycle parameters.

The user must supply a working directory to use for importing data and generating
outputs.

```{r packages, message=FALSE, warning=FALSE}
#----------------------- // USER defined variables //---------------------------
# define the working directory
working_directory <- "~/GitHub/SingleCellScripts/"

# run sctransform, this replaces the NormalizeData, ScaleData and FindVariableFeatures
sctransform = FALSE

# Should cell cycle filtering be applied or not?
filter_cell_cycle = FALSE
#--------------------------------------------------------------------------------------

# set other directories based on working directory
matrix_dir <- paste(working_directory,"data/count_matrices/",sep="")
rdata_dir <- paste(working_directory,"data/Rdata/",sep="")

# Check if directory for Rdata files exists, if not create it
ifelse(
  !dir.exists(rdata_dir), 
  dir.create(file.path(rdata_dir)), FALSE
)

# Store seurat object for quick re-use
seurat_object_save <- paste(rdata_dir,"seurat_object_preprocessed.Rdata",sep = "")

# Load packages
suppressMessages(library(tidyverse, quietly = TRUE))
suppressMessages(library(Seurat, quietly = TRUE))
```

## Load objects

```{r loadObjects, message=FALSE, warning=FALSE}
#---------------- scaffold cultures -------------------
# Analyze FUS-DDIT3 samples
fd_scf <- Read10X(paste(matrix_dir,"HT1080_Scf_FD/outs/filtered_gene_bc_matrices/custom_egfp_hg38/",sep=""))
fd_scf <- CreateSeuratObject(fd_scf,project = "scf_FD", min.cells = 5)
fd_scf$group <- "scf-FD"

# Analyze WT samples
wt_scf <- Read10X(paste(matrix_dir,"HT1080_Scf_WT/outs/filtered_gene_bc_matrices/custom_egfp_hg38/",sep=""))
wt_scf <- CreateSeuratObject(wt_scf,project = "scf_WT", min.cells = 5)
wt_scf$group <- "scf-WT"

#---------------- xenograft cultures -------------------
# Analyze WT samples
wt_xen <- Read10X(paste(matrix_dir,"HT1080_Xen_WT/outs/filtered_gene_bc_matrices/custom_egfp_hg38/",sep=""))
wt_xen <- CreateSeuratObject(wt_xen,project = "xen_WT", min.cells = 5)
wt_xen$group <- "xenograft-WT"

# Analyze FUS-DDIT3 samples
fd_xen <- Read10X(paste(matrix_dir,"HT1080_Xen_FD/outs/filtered_gene_bc_matrices/custom_egfp_hg38/",sep=""))
fd_xen <- CreateSeuratObject(fd_xen,project = "xen_FD", min.cells = 5)
fd_xen$group <- "xenograft-FD"
```

## Merge and save raw data

```{r merge, fig.width=8, fig.height=8, message=FALSE, warning=FALSE}
# Merged dataset
merged_data <- merge(
  x = wt_scf,
  y = c(wt_xen, fd_scf, fd_xen),
  add.cell.ids = c('wt-scf', 'wt-xen', 'fd-scf', 'fd-xen'),
  project = 'scaffold'
)

unique(x = sapply(X = strsplit(x = colnames(
  x = merged_data), split = '_'), FUN = '[', 1))

table(merged_data$orig.ident)

# store mitochondrial percentage in object meta data
merged_data[["percent.mt"]] <- PercentageFeatureSet(
  object = merged_data, 
  pattern = "^MT-"
)
```

### Quality control {.tabset}

Mitochondrial gene expression can be a confounding factor and very high
mtDNA expression might indicate dead cells which should be removed from further
analysis.

We plot nFeature_RNA (= number of genes per cell), nCount_RNA (= number of 
transcripts per cell) and percentage of mtDNA expression and filter cells
accordingly. When performing QC variables should be considered jointly.
For instance high mtDNa expression may also reflect cells with high 
respiratory activity rather than lysed cells.

```{r preproc, echo = FALSE}
plot1 <- FeatureScatter(
  object = merged_data,
  feature1 = "nCount_RNA",
  feature2 = "percent.mt",
  group.by = "group"
) + theme(
  legend.position = "bottom",
  legend.title = element_blank()
)

plot2 <- FeatureScatter(
  object = merged_data,
  feature1 = "nCount_RNA",
  feature2 = "nFeature_RNA",
  group.by = "group"
) + theme(
  legend.position = "bottom",
  legend.title = element_blank()
)
```

#### Violin

```{r, echo=FALSE}
VlnPlot(
  object = merged_data, 
  features = c('nFeature_RNA', 'nCount_RNA', 'percent.mt'), 
  ncol = 3, 
  group.by = 'group',
  log = TRUE
)
```

#### Scatter

```{r, echo =FALSE}
plot1 + plot2
```

### Filter dead cells and doublets {.tabset}

From the scatter plot of transcripts counts vs. mtDNA we can see that cells
with very high mtDNA content $>20\%$ also have very low transcript numbers and
are therefore probably lysed cells and should be removed.

```{r , message=FALSE}
# Filter data to remove dead cells, outliers
merged_data <- subset(
  x = merged_data,
  subset = nFeature_RNA > 2000 & nFeature_RNA < 8000 & percent.mt < 10
)

# Plot percent mito DNA and mRNA count after filtering
plot1 <- Seurat::FeatureScatter(
  object = merged_data,
  feature1 = "nCount_RNA",
  feature2 = "percent.mt",
  group.by = "group"
) + theme(
  legend.position = "bottom",
  legend.title = element_blank()
)

plot2 <- Seurat::FeatureScatter(
  object = merged_data,
  feature1 = "nCount_RNA",
  feature2 = "nFeature_RNA",
  group.by = "group"
) + theme(
  legend.position = "bottom",
  legend.title = element_blank()
)
```

#### Scatter
```{r , echo=FALSE}
plot1 + plot2
```

#### Violin

```{r }
# Visualize QC metrics as a violin plot
VlnPlot(
  object = merged_data, 
  features = c('nFeature_RNA', 'nCount_RNA', 'percent.mt'), 
  ncol = 3, 
  group.by = 'group'
)
```

### Cell cycle scoring

```{r cc_scoring, message=FALSE, warning=FALSE}
# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

# Assign cell cycle score to genes, which will be stored in the Seurat object
# metdata
merged_data <- Seurat::CellCycleScoring(
  object = merged_data,
  s.features = s.genes,
  g2m.features = g2m.genes,
  set.ident = TRUE
)

VlnPlot(
  object = merged_data, 
  features = c('nFeature_RNA', 'nCount_RNA', 'percent.mt'), 
  ncol = 3, 
  group.by = 'Phase'
)

VlnPlot(
  object = merged_data, 
  features = c('S.Score', 'G2M.Score'), 
  ncol = 3, 
  group.by = 'group'
)
```

### Normalizing the data using regularized negative binomial regression {.tabset}

We normalise the data either using [sctransform](https://www.biorxiv.org/content/10.1101/576827v1)
or using the standard Seurat workflow depending on the setting of the 
sctransform variable.

sctransform models the expression of each gene as a negative binomial random variable with a mean that depends on other variables. Here the other variables can be used to model the differences in sequencing depth between cells and are used as independent variables in a regression model. In order to avoid overfitting, we will first fit model parameters per gene, and then use the relationship between gene mean and parameter values to fit parameters, thereby combining information across genes. Given the fitted model parameters, we transform each observed UMI count into a Pearson residual which can be interpreted as the number of standard deviations an observed count was away from the expected mean. If the model accurately describes the mean-variance relationship and the dependency of mean and latent factors, then the result should have mean zero and a stable variance across the range of expression. During normalization, we can also remove confounding sources of variation, for example, mitochondrial mapping percentage.

We can also assign each cell a cell cycle score based on known cell-cycle
associated genes for S and G2/M phases. These are available through the Seurat
package.

We perform the normalization workflow and only regress out the variable 'percent.mt'.
We perform PCA on the scaled data with and withoutcell cycle regressed genes, however in both cases we regress out the mtDNAexpression. For this PCA we only use the annotated cell cycle genes! It is expected that cells cluster according to cell cycle stage without regression and that
regressed data shows much less separation according to cell cycle stage (but
not necessarily zero).

To filter cell cycle genes too set the filter_cell_cycle variable to TRUE.

This document will always show the effect of cell cycle regression by plotting the 
samples in PCA space with and without filtering, filter_cell_cycle only determines
which of the two will be used downstream.

```{r normalizeData, message=FALSE, warning=FALSE}
# Regress out mitochondrial expression and cell cycle stage
vars.to.regress = c('percent.mt', 'S.Score', 'G2M.Score')

# Decide which normalization workflow to use
if(sctransform){
  merged_data <- SCTransform(
    object = merged_data, 
    vars.to.regress = vars.to.regress,
    verbose = FALSE
  )
} else {
  merged_data <- Seurat::NormalizeData(
    object = merged_data,
    verbose = FALSE
  )
  
  merged_data <- Seurat::FindVariableFeatures(
    object = merged_data, 
    selection.method = 'vst', 
    nfeatures = 2000
  )
  
  # Scale data without regressing out cell cycle
  merged_data_cc <- Seurat::ScaleData(
    object = merged_data,
    vars.to.regress = 'percent.mt'
  )
  
  # Regress out cell cycle
  merged_data_no_cc <- Seurat::ScaleData(
    object = merged_data,
    vars.to.regress = vars.to.regress
  )
  
  # Decide to use data with/without cell cycle regression
  if(filter_cell_cycle){
    merged_data <- merged_data_no_cc
  } else {
    merged_data <- merged_data_cc
  }
  
  # Perform PCA on cc regressed samples
  merged_data_cc <- Seurat::RunPCA(
    object = merged_data_cc,
    features = c(s.genes, g2m.genes)
  )
  
  before_cc_correction <- Seurat::DimPlot(merged_data_cc)
  
  merged_data_no_cc <- Seurat::RunPCA(
    object = merged_data_no_cc,
    features = c(s.genes, g2m.genes)
  )
  
  after_cc_correction <- Seurat::DimPlot(merged_data_no_cc)
}
```

#### With cell cycle
```{r, echo=FALSE, message=FALSE}
before_cc_correction
```

#### Cell cycle regressed
```{r, echo=FALSE, message=FALSE}
after_cc_correction
```

## Save raw data

```{r}
save(merged_data, file = seurat_object_save)
```

