---
title: "Single-cell RNA seq  trajctories and pseudotime plots"
author: "Stefan Filges"
date: '`r format(Sys.Date())`'
output: 
  html_document:
    theme: sandstone
    highlight: tango
    code_folding: hide
    toc: true
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load_libraries}
suppressMessages(library(Seurat))
suppressMessages(library(monocle))
suppressMessages(library(viridis))
suppressMessages(library(RaceID))
suppressMessages(library(tidyverse))
suppressMessages(library(RColorBrewer))
suppressMessages(library(magrittr))
```

# What does this document do?

This document uses pre-generated single cell data objects containing pseudotime/trajctory information and generates plots for data exploration and analysis. The data sets were generated using a different script, which can take a significant amount of time and is a long document containing code on many other issues. This document allows for faster and easier-to-use analysis.

# Monocle 2

## Load data

First we load a CellDataSet called cds generated with Monocle v2. This contains the expression data, cell metdata (phenotypes) as well as pseudotime values.

```{r import_data}
load(file = './Data/cds_monocle2.Rdata')
load(file = './Data/seurat_object.Rdata')

clustering_DEG_genes = read.csv2(file = './Data/monocle2_DEG_genes.csv')
order_genes <- read.csv2(file = './Data/ordering_genes.csv')
```

## Cell trajectories

```{r monocle_v2_pseudotime, echo=FALSE}
plot1 <- monocle::plot_cell_trajectory(
  cds = cds,
  color_by = "Pseudotime"
) + scale_color_viridis_c()

plot2 <- monocle::plot_cell_trajectory(
  cds = cds, 
  color_by = "group"
) + scale_color_manual(
  values = c('#ff8668', 'gray70', '#5e8aab', '#ffce8e')
  )

#+ scale_color_brewer(palette = "Dark2")
#+ scale_color_manual("#ff8668","gray70","#5e8aab","#ffce8e")

plot1 + plot2
```

## Expression heatmap

We want to find genes differentially expressed across pseudotime. Instead of
testing all genes, we will limit the number of genes to be tested: (i) First
we select only genes differentially expressed in cluster with an adjusted
p value $<0.01$ and (ii) select the top 500 of those to perform the test. 

After testing we select again based on adjusted p value $<0.01$ to find the 
most significantly expressed genes along the trajectory. These are then visualised
in a heatmap with 

```{r monocle2-heatmap, eval=TRUE, message=FALSE}
clustering_genes <- clustering_DEG_genes %>% 
  dplyr::filter(use_for_ordering == TRUE) %>%
  dplyr::arrange(qval)

to_be_tested <- clustering_genes$gene_short_name[1:1500]
#to_be_tested <- order_genes$gene

cds_subset <- cds[as.character(to_be_tested),]

diff_test_res <- monocle::differentialGeneTest(
  cds_subset,
  fullModelFormulaStr = "~sm.ns(Pseudotime)"
)

sig_gene_names <- row.names(subset(diff_test_res, qval < 0.01))

heatmap <- monocle::plot_pseudotime_heatmap(
  cds_subset[sig_gene_names,],
  num_clusters = 4,
  cores = 4,
  cluster_rows = TRUE,
  norm_method = 'vstExprs',
  show_rownames = FALSE,
  hmcols = viridis::magma(256),
  return_heatmap = TRUE
)

pdf(file = './Data/heatmap_4_clusters.pdf', width = 7,height = 3)
heatmap
dev.off()

# Retrieve gene names + cluster identities
clusters <- cutree(heatmap$tree_row, k = 4)
t <- tibble(
  gene = names(clusters), 
  cluster = clusters
  ) %>%
  arrange(cluster)

readr::write_csv2(t, path = './Data/diff_genes_over_pseudotime_top_1500_q_0.01.csv')
```

```{r}
cds_time <- cds_subset[,order(cds_subset$Pseudotime)]

# red: scf FD - #F8766D
# grÃ¼n: scf WT - #7CAE00
# cyan: xen FD - #00BFC4
# violett: xen WT - #C77CFF

ggplotColours <- function(n = 6, h = c(0, 360) + 15){
  if ((diff(h) %% 360) < 1) h[2] <- h[2] - 360/n
  hcl(h = (seq(h[1], h[2], length = n)), c = 100, l = 65)
}

annotation = list(
  time = cds_time$Pseudotime,
  group = cds_time$group
)
annotation_colors = list(
  time = viridis::viridis(256),
  group =  c("#ff8668", "gray70","#5e8aab","#ffce8e")
)

NMF::aheatmap(
  x = as.matrix(exprs(cds_time[1:10,])),
  Colv = NA, 
  Rowv = NA, 
  border_color = NA,
  annCol = annotation,
  annColors = annotation_colors,
  scale = 'none',
  width = 5,
  height = 5,
  filename = "aheatmap_500_genes.pdf"
)
```

## Plot differentially expressed genes in pseudotime

```{r}
diff_test_res <- diff_test_res %>% 
  dplyr::filter(qval < 0.01) %>%
  dplyr::arrange(qval)

my_genes <- c('CDKN2C', 'LAMA4', 'GAPDH', 'APOE', 'ALDOA')
cds_subset <- cds[my_genes,]

#markers_to_plot <- t %>% dplyr::group_by(cluster) %>% summarise(first = first(gene)) 
#cds_subset <- cds[markers_to_plot$first,]

# Plot by pseudotime
plot1 <- monocle::plot_genes_in_pseudotime(
  cds_subset = cds_subset,
  color_by = "Pseudotime",
  relative_expr = TRUE,
  cell_size = 0.5
) + scale_color_viridis_c() +
  theme(legend.position = 'bottom')

# Plot by sample
plot2 <- monocle::plot_genes_in_pseudotime(
  cds_subset = cds_subset,
  relative_expr = TRUE,
  color_by = "group",
  cell_size = 0.5
  ) + scale_color_manual(values = c("#ff8668", "gray70","#5e8aab","#ffce8e"))+
  theme(legend.position = 'bottom')

#+ scale_color_brewer(palette = "Dark2")
#+ scale_color_manual(values = c("#ff8668", "gray70","#5e8aab","#ffce8e"))

plot1 + plot2
```

## RaceID

```{r, raceID, eval=FALSE}
# Create the sc object
ndata <- merged_data@assays$RNA@counts

sc <- RaceID::SCseq(ndata)

# Filter for this
sc <- RaceID::filterdata(
  object = sc,
  mintotal = 2000
)

# Regress variables from cell object
vars <- data.frame(
  nCount_RNA = merged_data$nCount_RNA,
  S.score = merged_data$S.Score,
  G2M.score = merged_data$G2M.Score
)
sc <- RaceID::varRegression(sc,vars)

# Re-initialize raceID
part <- as.numeric(merged_data@meta.data$seurat_clusters)
d <- as.matrix(dist(merged_data@reductions$pca@cell.embeddings))
umap <- as.data.frame(merged_data@reductions$umap@cell.embeddings)
names(part) <- colnames(d)

n <- colnames(sc@ndata)
part <- part[n]

# partition
sc@cpart <- sc@cluster$kpart <- part
# distances
sc@distances <- d[n,n]
# umap
sc@umap <- umap[n,]
# expression data (optional)
sc@counts <- sc@counts * 0 + 1
sc@ndata  <- ndata[,n]
# cluster medoids
sc@medoids <- RaceID::compmedoids(sc, sc@cpart)

col_cluster <- colorRampPalette(
  brewer.pal(12,"Set3"))(
    length(
      unique(
        merged_data@meta.data$seurat_clusters)
      )
    )

names(col_cluster) <- as.character(unique(merged_data@meta.data$seurat_clusters))

set.seed(12345)
sc@fcol <- col_cluster[order(as.numeric(names(col_cluster)))]

# Run StemID
# Initialise lineage tree object
ltr <- RaceID::Ltree(sc)
# Compute transcriptome entropy
ltr <- RaceID::compentropy(ltr)
# Project cells unto cluster links
ltr <- RaceID::projcells(
  object = ltr,
  cthr=5,     # use clusters with at least this many cells
  nmode=TRUE, # use knn
  um=TRUE,    # use UMAP projection for visualisation
  knn=3
)

ltr <- RaceID::lineagegraph(object = ltr)

ltr <- RaceID::comppvalue(ltr, pthr=0.1)

RaceID::plotgraph(ltr,showCells=FALSE,showMap=TRUE)

x <- RaceID::compscore(ltr)

# Plot stemID score for each cluster
stem_data <- tibble(
  cluster = names(x$StemIDscore), 
  score = x$StemIDscore
  ) %>%
  dplyr::arrange(score)

stem_data$cluster <- as.factor(str_remove(stem_data$cluster, "cl.")) 
stem_data$color <- fct_relevel(stem_data$cluster, c("10", "11", "12", "13"), 
  after = Inf)

stem_score <- ggplot(
  data = stem_data,
  mapping = aes(x = reorder(cluster, score), y=score, fill = color)
  ) + 
  geom_bar(
    stat = "identity",
    colour="black",
    width = .8,
    size = .01
    ) + 
  theme_bw()

stem_score
```


