---
title: "(5) Trajctories and pseudotime plots"
author: "Stefan Filges"
date: '`r format(Sys.Date())`'
output: 
  html_document:
    theme: sandstone
    highlight: tango
    code_folding: show
    toc: yes
    toc_depth: 3
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load_libraries}
# Packages for main analysis, clustering, data structures etc.
suppressMessages(library(Seurat, quietly = TRUE))
suppressMessages(library(biomaRt, quietly = TRUE))
suppressMessages(library(msigdbr, quietly = TRUE))
suppressMessages(library(clusterProfiler, quietly = TRUE))
suppressMessages(library(SummarizedExperiment, quietly = TRUE))
suppressMessages(require(clusterExperiment, quietly = TRUE))

# Packages for data handeling and plotting
suppressMessages(library(tidyverse, quietly = TRUE))
suppressMessages(library(viridis, quietly = TRUE))
suppressMessages(library(RColorBrewer, quietly = TRUE))

# Packages for pseudotime ordering
suppressMessages(library(slingshot, quietly = TRUE))
suppressMessages(library(monocle, quietly = TRUE))
suppressMessages(library(SCORPIUS, quietly = TRUE))
```

# What does this document do?

This document uses pre-generated single cell data objects containing pseudotime/trajctory information and generates plots for data exploration and analysis. The data sets were generated using a different script, which can take a significant amount of time and is a long document containing code on many other issues. This document allows for faster and easier-to-use analysis.

# Monocle 2

## Load data

First we load a CellDataSet called cds generated with Monocle v2. This contains the expression data, cell metdata (phenotypes) as well as pseudotime values.

```{r import_data}
# define the working directory
working_directory <- "~/GitHub/SingleCellScripts/"

# set other directories based on working directory
figures_dir <- paste(working_directory,"figures/",sep="")
rdata_dir <- paste(working_directory,"data/Rdata/",sep="")

# load object
load(file = paste(rdata_dir, 'seurat_object_final.Rdata', sep = ''))
load(file = paste(rdata_dir, 'cds_monocle2.Rdata', sep = ''))

source(paste(working_directory, 'R/helper_functions.R', sep = ''))

clustering_DEG_genes = read.csv2(file = paste(working_directory, 'data/gene_lists/monocle2_DEG_genes.csv', sep = ''))
order_genes <- read.csv2(file = paste(working_directory, 'data/gene_lists/ordering_genes.csv', sep = ''))
```

## Cell trajectories

```{r monocle_v2_pseudotime, echo=FALSE}
plot1 <- monocle::plot_cell_trajectory(
  cds = cds,
  color_by = "Pseudotime"
) + scale_color_viridis_c()

plot2 <- monocle::plot_cell_trajectory(
  cds = cds, 
  color_by = "group"
) + scale_color_manual(values = c("#ff8668", "gray70","#5e8aab","#ffce8e"))

# Save monocle pseudotime
ggplot2::ggsave(filename = '~/GitHub/SingleCellScripts/figures/monocle_pseudotime.svg', plot = plot1, width = 6, height = 6, device = "svg")
ggplot2::ggsave(filename = '~/GitHub/SingleCellScripts/figures/monocle_pseudotime.pdf', plot = plot1, width = 6, height = 6, device = "pdf")

# Save moncole samples
ggplot2::ggsave(filename = '~/GitHub/SingleCellScripts/figures/monocle_samples.svg', plot = plot2, width = 6, height = 6, device = "svg")
ggplot2::ggsave(filename = '~/GitHub/SingleCellScripts/figures/monocle_samples.pdf', plot = plot2, width = 6, height = 6, device = "pdf")

plot1 + plot2
```

## Expression heatmap

We want to find genes differentially expressed across pseudotime. Instead of
testing all genes, we will limit the number of genes to be tested: (i) First
we select only genes differentially expressed in cluster with an adjusted
p value $<0.01$ and (ii) select the top 500 of those to perform the test. 

After testing we select again based on adjusted p value $<0.01$ to find the 
most significantly expressed genes along the trajectory. These are then visualised
in a heatmap with 

```{r monocle2-heatmap, eval=TRUE, message=FALSE}
clustering_genes <- clustering_DEG_genes %>% 
  dplyr::filter(use_for_ordering == TRUE) %>%
  dplyr::arrange(qval)

to_be_tested <- clustering_genes$gene_short_name[1:1500]

cds_subset <- cds[as.character(to_be_tested),]

diff_test_res <- monocle::differentialGeneTest(
  cds_subset,
  fullModelFormulaStr = "~sm.ns(Pseudotime)"
)

sig_gene_names <- row.names(subset(diff_test_res, qval < 0.01))

#heatmap <- monocle::plot_pseudotime_heatmap(
#  cds_subset[sig_gene_names,],
#  num_clusters = 4,
#  cores = 4,
#  cluster_rows = TRUE,
#  norm_method = 'vstExprs',
#  show_rownames = FALSE,
#  hmcols = viridis::magma(256),
#  return_heatmap = TRUE
#)

heatmap <- pseudotime_hm(
  cds_subset[sig_gene_names,],
  num_clusters = 4,
  cores = 4,
  cluster_rows = TRUE,
  norm_method = 'vstExprs',
  show_rownames = FALSE,
  hmcols = viridis::magma(256),
  return_heatmap = TRUE
)


#pdf(file = '~/GitHub/SingleCellScripts/figures/heatmap_4_clusters.pdf', width = 7,height = 3)
#heatmap
#dev.off()

# Save cluster1 enrichments
ggplot2::ggsave(filename = paste(figures_dir, 'pseudotime_heatmap.svg', sep = ''), plot = heatmap, width = 7, height = 3, device = "svg")
ggplot2::ggsave(filename = paste(figures_dir, 'pseudotime_heatmap.pdf', sep = ''), plot = heatmap, width = 7, height = 3, device = "pdf")

# Retrieve gene names + cluster identities
clusters <- cutree(heatmap$tree_row, k = 4)
t <- tibble(
  gene = names(clusters), 
  cluster = clusters
  ) %>%
  arrange(cluster)

readr::write_csv2(t, path = '~/GitHub/SingleCellScripts/data/gene_lists/diff_genes_over_pseudotime_top_1500_q_0.01.csv')
```

```{r, eval = FALSE}
cds_time <- cds_subset[,order(cds_subset$Pseudotime)]

# red: scf FD - #F8766D
# grÃ¼n: scf WT - #7CAE00
# cyan: xen FD - #00BFC4
# violett: xen WT - #C77CFF

ggplotColours <- function(n = 6, h = c(0, 360) + 15){
  if ((diff(h) %% 360) < 1) h[2] <- h[2] - 360/n
  hcl(h = (seq(h[1], h[2], length = n)), c = 100, l = 65)
}

annotation = list(
  time = cds_time$Pseudotime,
  group = cds_time$group
)
annotation_colors = list(
  time = viridis::viridis(256),
  group =  c("#ff8668", "gray70","#5e8aab","#ffce8e")
)

# c("#003f5c","#7a5195","#ef5675","#ffa600")
# c("#ff8668", "gray70","#5e8aab","#ffce8e")

NMF::aheatmap(
  x = as.matrix(exprs(cds_time[1:10,])),
  Colv = NA, 
  Rowv = NA, 
  border_color = NA,
  annCol = annotation,
  annColors = annotation_colors,
  scale = 'none',
  width = 5,
  height = 5,
  filename = "~/GitHub/SingleCellScripts/figures/aheatmap_1500_genes.pdf"
)
```

## Plot differentially expressed genes in pseudotime

```{r, eval = FALSE}
diff_test_res <- diff_test_res %>% 
  dplyr::filter(qval < 0.01) %>%
  dplyr::arrange(qval)

#my_genes <- c('IL6', 'ENO2', 'IL24', 'KRT80', 'MME')
#cds_subset <- cds[my_genes,]

# Methylation associated targets
HDAC1_TARGETS_UP <- read_delim(paste(working_directory,'data/MSigDB/SENESE_HDAC1_TARGETS_UP.txt',sep=''), delim = ' ')
HDAC3_TARGETS_UP <- read_delim(paste(working_directory,'data/MSigDB/SENESE_HDAC3_TARGETS_UP.txt',sep=''), delim = ' ')
SATB1_TARGETS_UP <- read_delim(paste(working_directory,'data/MSigDB/HAN_SATB1_TARGETS_UP.txt',sep=''), delim = ' ')
SILENCED_BY_METHYLATION_UP <- read_delim(paste(working_directory,'data/MSigDB/HELLER_HDAC_TARGETS_SILENCED_BY_METHYLATION_UP.txt',sep=''), delim = ' ')



my_genes <- t %>% dplyr::filter(cluster == 2)
my_genes <- my_genes[my_genes$gene %in% SILENCED_BY_METHYLATION_UP$HELLER_HDAC_TARGETS_SILENCED_BY_METHYLATION_UP,]
cds_subset <- cds[my_genes$gene[1:5],]


#my_genes <- c('IL24', 'SERPINE1', 'GRN', 'NAMPT')
#my_genes <- c('TGFB1', 'CXCL8', 'IL1B', 'FLT1', 'IL6R')
my_genes <- c('CD44', 'SERPINB8', 'SLC16A1', 'CEBPD')
#my_genes <- c('CLU', 'OSMR', 'STAT1', 'BIN1')
#my_genes <- c('IL6', 'PIAS3', 'STAT6', 'IFI6', 'IFI27')


my_genes <- c('CD274','IFI44', 'OASL', 'MBD1', 'TRIM22')


my_genes <- c('ATRX')
cds_subset <- cds[my_genes,]

# Plot by pseudotime
plot1 <- monocle::plot_genes_in_pseudotime(
  cds_subset = cds_subset,
  color_by = "Pseudotime",
  relative_expr = TRUE,
  cell_size = 0.5
) + scale_color_viridis_c() +
  theme(legend.position = 'bottom')

# Plot by sample
plot2 <- monocle::plot_genes_in_pseudotime(
  cds_subset = cds_subset,
  relative_expr = TRUE,
  color_by = "group",
  cell_size = 0.5
  ) + scale_color_manual(values = c("#ff8668", "gray70","#5e8aab","#ffce8e")) +
  theme(legend.position = 'bottom')

#+ scale_color_brewer(palette = "Dark2")
#+ scale_color_manual(values = c("#ff8668", "gray70","#5e8aab","#ffce8e"))

plot1 + plot2
```

```{r, eval = TRUE}

#my_genes <- c('IL24', 'SERPINE1', 'GRN', 'NAMPT')
#my_genes <- c('TGFB1', 'CXCL8', 'IL1B', 'FLT1', 'IL6R')
#my_genes <- c('CD44', 'SERPINB8', 'SLC16A1', 'CEBPD')
#my_genes <- c('CLU', 'OSMR', 'STAT1', 'BIN1')
#my_genes <- c('IL6', 'PIAS3', 'STAT6', 'IFI6', 'IFI27')
#my_genes <- c('CD274','IFI44', 'OASL', 'MBD1', 'TRIM22')

my_genes <- c('IL24', 'SERPINE1', 'GRN', 'NAMPT', 'TGFB1', 'CXCL8', 'IL1B', 
              'FLT1', 'IL6R','CD44', 'SERPINB8', 'SLC16A1', 'CEBPD', 'CLU', 
              'OSMR', 'STAT1', 'BIN1', 'IL6', 'PIAS3', 'STAT6', 'IFI6', 
              'IFI27', 'CD274','IFI44', 'OASL', 'MBD1', 'TRIM22', 'KDM5A', 
              'RUNX2', 'EPAS1', 'HDAC1', 'HDAC2', 'HDAC3', 'HDAC4', 'HDAC5', 
              'HDAC6', 'HDAC7', 'HDAC8', 'HDAC9', 'STAT2', 'STAT3', 'STAT5A', 
              'STAT5B','PIAS1', 'PIAS2', 'PIAS3', 'PIAS4', 'EZH2', 'SUZ12',
              'SMARCE1', 'SMARCA5', 'SMARCD1', 'ACTL6A')


for(gene in my_genes){
  
  fplot <- Seurat::FeaturePlot(
    object = merged_data,
    features = gene
  )
  
  vln_plot <- Seurat::VlnPlot(
    object = merged_data,
    features = gene,
    slot = "counts",
    log = TRUE,
    group.by = "group"
  )

  ggplot2::ggsave(
    filename = paste(figures_dir,'feature_plots/', gene, '_feature_plot.png', sep = ''),
    plot = fplot, 
    width = 6, 
    height = 6, 
    device = "png"
  )
  
  ggplot2::ggsave(
    filename = paste(figures_dir,'feature_plots/', gene, '_violin_plot.png', sep = ''),
    plot = vln_plot, 
    width = 6, 
    height = 6, 
    device = "png"
  )
}
```


```{r, eval = FALSE}
cluster_1 <- enrichment_plot(data = t, id = 1)
cluster_2 <- enrichment_plot(data = t, id = 2)
cluster_3 <- enrichment_plot(data = t, id = 3)
cluster_4 <- enrichment_plot(data = t, id = 4)

# Save cluster1 enrichments
ggplot2::ggsave(filename = paste(figures_dir, 'enrichment_cluster_1.svg', sep = ''), plot = cluster_1, width = 12, height = 4, device = "svg")
ggplot2::ggsave(filename = paste(figures_dir, 'enrichment_cluster_1.pdf', sep = ''), plot = cluster_1, width = 12, height = 4, device = "pdf")

# Save cluster2 enrichments
ggplot2::ggsave(filename = paste(figures_dir, 'enrichment_cluster_2.svg', sep = ''), plot = cluster_2, width = 12, height = 4, device = "svg")
ggplot2::ggsave(filename = paste(figures_dir, 'enrichment_cluster_2.pdf', sep = ''), plot = cluster_2, width = 12, height = 4, device = "pdf")

# Save cluster3 enrichments
ggplot2::ggsave(filename = paste(figures_dir, 'enrichment_cluster_3.svg', sep = ''), plot = cluster_3, width = 12, height = 4, device = "svg")
ggplot2::ggsave(filename = paste(figures_dir, 'enrichment_cluster_3.pdf', sep = ''), plot = cluster_3, width = 12, height = 4, device = "pdf")

# Save cluster4 enrichments
ggplot2::ggsave(filename = paste(figures_dir, 'enrichment_cluster_4.svg', sep = ''), plot = cluster_4, width = 12, height = 4, device = "svg")
ggplot2::ggsave(filename = paste(figures_dir, 'enrichment_cluster_4.pdf', sep = ''), plot = cluster_4, width = 12, height = 4, device = "pdf")

```

## Moncole 3

```{r moncole3, eval = FALSE}
library(monocle3)
library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(ggplot2)
library(patchwork)
library(magrittr)

integrated <- ScaleData(merged_data)
integrated <- RunPCA(integrated)
integrated <- RunUMAP(integrated, dims = 1:30, reduction.name = "UMAP")
integrated <- FindNeighbors(integrated, dims = 1:30)
integrated <- FindClusters(integrated)
DimPlot(integrated, group.by = c("orig.ident", "ident"))


cds <- as.cell_data_set(integrated)
cds <- cluster_cells(cds)
cds <- learn_graph(cds,use_partition = FALSE)
cds <- order_cells(cds)


p1 <- plot_cells(cds, color_cells_by = "pseudotime", label_cell_groups = FALSE, label_leaves = FALSE, 
    label_branch_points = FALSE)
p2 <- plot_cells(cds, color_cells_by = "group", show_trajectory_graph = TRUE)
wrap_plots(p1, p2)






```

